cs3api4lab/cs3_file_api.py

```diff
@@ -0,0 +1,237 @@
+            auth_res = self.cs3_stub.Authenticate(auth_req)
+            self.tokens[userid] = {'tok': auth_res.token, 'exp': time.time() + self.authtokenvalidity}
+
+        # piggy back on the opportunity to expire old tokens, but asynchronously
+        # as to not impact the current session: let's use python3.7's coroutines support
+        # asyncio.run(_async_cleanuptokens())
+
+        return self.tokens[userid]['tok']
+
+    def _cs3_reference(self, endpoint, fileid):
+
+        if fileid[0] == '/':
+            # assume this is a filepath
+            return cs3spr.Reference(path=fileid)
+
+        if endpoint == 'default' or endpoint is None:
+            raise IOError('A CS3API-compatible storage endpoint must be identified by a storage UUID')
+
+        # assume we have an opaque fileid
+        return cs3spr.Reference(id=cs3spr.ResourceId(storage_id=endpoint, opaque_id=fileid))
+
+    def stat(self, endpoint, fileid, userid):
+
+        """
+        Stat a file and returns (size, mtime) as well as other extended info using the given userid as access token.
+        Note that endpoint here means the storage id. Note that fileid can be either a path (which MUST begin with /), or an id (which MUST NOT
+        start with a /).
+        """
+
+        time_start = time.time()
+
+        ref = self._cs3_reference(endpoint, fileid)
+
+        stat_info = self.cs3_stub.Stat(request=cs3sp.StatRequest(ref=ref), metadata=[('x-access-token', self._authenticate(userid))])
+
+        time_end = time.time()
+        self.log.info('msg="Invoked stat" fileid="%s" elapsedTimems="%.1f"' % (fileid, (time_end - time_start) * 1000))
+
+        if stat_info.status.code == cs3code.CODE_OK:
+            self.log.debug('msg="Stat result" data="%s"' % stat_info)
+            return {
+                'inode': stat_info.info.id.storage_id + ':' + stat_info.info.id.opaque_id,
+                'filepath': stat_info.info.path,
+                'userid': stat_info.info.owner.opaque_id,
+                'size': stat_info.info.size,
+                'mtime': stat_info.info.mtime.seconds
+            }
+
+        self.log.info('msg="Failed stat" fileid="%s" reason="%s"' % (fileid, stat_info.status.message))
+        raise IOError(stat_info.status.message)
+
+    def read_file(self, endpoint, filepath, userid):
+        """
+        Read a file using the given userid as access token.
+        """
+
+        time_start = time.time()
+
+        #
+        # Prepare endpoint
+        #
+        reference = self._cs3_reference(endpoint, filepath)
+
+        req = cs3sp.InitiateFileDownloadRequest(ref=reference)
+
+        init_file_download = self.cs3_stub.InitiateFileDownload(request=req, metadata=[('x-access-token', self._authenticate(userid))])
+
+        if init_file_download.status.code == cs3code.CODE_NOT_FOUND:
+            self.log.info('msg="File not found on read" filepath="%s"' % filepath)
+            yield IOError('No such file or directory')
```

Comment:
Why do we yield instead of raise? If this fails I don't see why we would continue.
Or do you have any reason?
(check the others)

Reply:
It's form WOPI example. It using yeld if have an error in read file at async, but we can change to "raise".